axes/handlers/database.py

```diff
@@ -219,6 +220,21 @@ def user_login_failed(self, sender, credentials: dict, request=None, **kwargs):
                         client_str,
                     )
 
+                if settings.AXES_USE_ATTEMPT_EXPIRATION:
+                    if not hasattr(attempt, "expiration") or attempt.expiration is None:
```

Comment:
Maybe just create the object if it doesn't exist? 

Is the expiration timestamp ever *earlier* than it previously was? Why should it only be updated to a later value if the `get_attempt_expiration` returns an earlier value?

```suggestion
                if settings.AXES_USE_ATTEMPT_EXPIRATION:
                    expiration, created = AccessAttemptExpiration.objects.update_or_create(
                            access_attempt=attempt,
                            defaults={
                                "expires_at": get_attempt_expiration(request),
                            }
                        )
                    if created:
                        log.debug("AXES: Created new AccessAttemptExpiration for %s", client_str)
                    else:
                        log.debug("AXES: Updated existing AccessAttemptExpiration for %s", client_str)           
```